frontend/src/features/admin-form/responses/ChartsPage/UnlockedCharts/UnlockedChartsContainer.tsx

```diff
@@ -0,0 +1,224 @@
+
+import { DecryptedSubmission } from '../../AdminSubmissionsService'
+import { useStorageResponsesContext } from '../../ResponsesPage/storage'
+import { useAllSubmissionData } from '../queries'
+
+import { EmptyChartsContainer } from './components/EmptyChartsContainer'
+import { FIELD_TO_CHART, FormChart } from './components/FormChart'
+import WordCloud, { WordCloudProps } from './components/WordCloud'
+
+// transform filtered data into an array of answer to count
+const aggregateSubmissionData = (
+  id: string,
+  formField: FormFieldDto,
+  data: DecryptedSubmission[],
+): [string, number][] => {
+  const hashMap = new Map<string, number>()
+  if (formField.fieldType === BasicField.Rating) {
+    for (let i = 1; i <= formField.ratingOptions.steps; i += 1) {
+      hashMap.set(String(i), 0)
+    }
+  }
+
+  data.forEach((content) => {
+    content.responses.forEach((field) => {
+      if (field._id === id && field.answer) {
+        // singular answer fields
+        hashMap.set(field.answer, (hashMap.get(field.answer) || 0) + 1)
+      } else if (field._id === id && field.answerArray) {
+        // multi answer fields, like checkboxes
+        field.answerArray.forEach((answer) => {
+          if (typeof answer === 'string')
+            return hashMap.set(answer, (hashMap.get(answer) || 0) + 1)
+        })
+      }
+    })
+  })
+
+  return Array.from(hashMap)
+}
+
+// transform filtered text data into an array of {word: count}
+const aggregateWordCloud = (
+  id: string,
+  data: DecryptedSubmission[],
+): WordCloudProps['words'] => {
+  const hashMap = new Map<string, number>()
+
+  const resultArr: WordCloudProps['words'] = []
+
+  data.forEach((content) => {
+    content.responses.forEach((field) => {
+      if (field._id === id && field.answer) {
+        // split to words
+        const answerArray = field.answer.split(' ')
+        // remove stop words from array
+        const ansNoStopW = removeStopwords(answerArray)
+        ansNoStopW.forEach((word) => {
+          // remove punctuations
+          const wordNoPunc = word.replace(/\W|_/g, '')
+          // normalise to lower case
+          const wordLower = wordNoPunc.toLowerCase()
+          hashMap.set(wordLower, (hashMap.get(wordLower) || 0) + 1)
+        })
+      }
+    })
+  })
+  hashMap.forEach((val, key) => resultArr.push({ text: key, value: val }))
+  return resultArr
+}
+
+export const UnlockedChartsContainer = () => {
+  const { data: decryptedContent } = useAllSubmissionData()
+  const { data: form } = useAdminForm()
+  const { dateRange, setDateRange } = useStorageResponsesContext()
+
+  const filteredDecryptedData: DecryptedSubmission[] = useMemo(() => {
+    if (!decryptedContent) return []
+    if (dateRange.length === 2) {
+      const [startDate, endDate] = dateRange.map((date) =>
+        new Date(date).getTime(),
+      )
+      // Set to the end of the endDate
+      const endOfDay = addDays(endDate, 1).getTime() - 1
```

Comment:
might have clearner way to do this

Reply:
e586558a8 the previous implementation has a bug where it adds 24 hours, and removes 1 ms from it, instead of the actual end of current day.